File: interpark_checker.py
import requests
import xml.etree.ElementTree as ET
import json
import time
import random
import logging
from pathlib import Path

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s  %(levelname)s  %(message)s",
    datefmt="%Y-%m-%d %H:%M:%S",
)
log = logging.getLogger(__name__)

SLACK_CONFIG_FILE = Path("slack_config.json")
TARGET_FILE = Path("interpark_target.json")

API_URL = "https://poticket.interpark.com/Book/Lib/BookInfoXml.asp"

HEADERS = {
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/147.0.0.0 Safari/537.36",
    "Accept": "application/xml, text/xml, */*",
    "Accept-Language": "ko-KR,ko;q=0.9,en-US;q=0.8,en;q=0.7",
    "Referer": "https://poticket.interpark.com/",
}


def load_slack_config():
    if not SLACK_CONFIG_FILE.exists():
        raise FileNotFoundError("slack_config.json 없음.")
    cfg = json.loads(SLACK_CONFIG_FILE.read_text(encoding="utf-8"))
    if not cfg.get("webhook_url"):
        raise ValueError("slack_config.json에 webhook_url을 입력해주세요.")
    return cfg


def load_target():
    if not TARGET_FILE.exists():
        raise FileNotFoundError("interpark_target.json 없음.")
    data = json.loads(TARGET_FILE.read_text(encoding="utf-8"))
    for key in ("GoodsCode", "PlaceCode", "PlaySeq", "SessionId"):
        if not data.get(key):
            raise ValueError(f"interpark_target.json에 {key} 값이 없습니다.")
    return data


def parse_grades(xml_text):
    root = ET.fromstring(xml_text)
    grades = []
    for table in root.findall("Table"):
        grades.append({
            "grade": table.findtext("SeatGrade"),
            "name": table.findtext("SeatGradeName"),
            "remain": int(table.findtext("RemainCnt") or 0),
            "price": table.findtext("SalesPrice"),
        })
    return grades


def send_slack(webhook_url, message):
    try:
        r = requests.post(webhook_url, json={"text": message}, timeout=10)
        r.raise_for_status()
        log.info("슬랙 알림 전송 완료")
    except Exception as e:
        log.error(f"슬랙 전송 실패: {e}")


def check_ticket(target, webhook_url):
    params = {
        "Flag": "OrderSeatGrade",
        "GoodsCode": target["GoodsCode"],
        "PlaceCode": target["PlaceCode"],
        "BizCode": target.get("BizCode", "WEBBR"),
        "PlaySeq": target["PlaySeq"],
        "SessionId": target["SessionId"],
        "InterlockingGoods": "",
    }

    session = requests.Session()
    session.cookies.update(target.get("cookies", {}))

    try:
        resp = session.get(API_URL, params=params, headers=HEADERS, timeout=15)
        resp.raise_for_status()
        resp.encoding = "utf-8"

        grades = parse_grades(resp.text)

        # 감시 등급 필터 (설정 없으면 전체 감시)
        watch_grades = target.get("watch_grades")

        available = [
            g for g in grades
            if g["remain"] > 0 and (not watch_grades or g["name"] in watch_grades)
        ]

        status = " | ".join(f'{g["name"]} {g["remain"]}매' for g in grades)
        log.info(status)

        if available:
            lines = [f'*{g["name"]}* {g["remain"]}매 ({int(g["price"]):,}원)' for g in available]
            play_date = target.get("PlayDate", "")
            date_str = f"{play_date[:4]}.{play_date[4:6]}.{play_date[6:]}" if play_date else ""
            goods_name = target.get("GoodsName", "")
            send_slack(
                webhook_url,
                f"🎫 인터파크 취소표 발생!\n"
                + (f"*{goods_name}*\n" if goods_name else "")
                + (f"📅 {date_str} {target.get('PlaySeq', '')}회차\n" if date_str else "")
                + "\n".join(lines)
                + f"\n🔗 https://ticket.interpark.com/Ticket/Goods/GoodsInfo.asp?GoodsCode={target['GoodsCode']}",
            )
            return "available"

        return "none"

    except ET.ParseError as e:
        log.error(f"XML 파싱 실패 (세션 만료 가능성): {e}")
        send_slack(
            webhook_url,
            "⚠️ 인터파크 응답 파싱 실패.\n"
            "interpark_target.json의 SessionId 또는 cookies를 갱신해주세요.",
        )
        return "session_expired"

    except Exception as e:
        log.error(f"API 호출 실패: {e}")
        return "error"


def main():
    slack = load_slack_config()
    webhook_url = slack["webhook_url"]

    log.info("인터파크 티켓 체커 시작 (20~40초 랜덤 간격)")
    send_slack(webhook_url, "🔍 인터파크 취소표 모니터링 시작")

    consecutive_errors = 0

    while True:
        try:
            target = load_target()
        except (FileNotFoundError, ValueError) as e:
            log.error(str(e))
            break

        result = check_ticket(target, webhook_url)

        if result == "error":
            consecutive_errors += 1
            if consecutive_errors >= 5:
                send_slack(
                    webhook_url,
                    f"❌ {consecutive_errors}회 연속 오류. 네트워크 또는 API 문제를 확인해주세요.",
                )
        else:
            consecutive_errors = 0

        if result == "session_expired":
            log.info("세션 만료 — 60초 대기 후 재시도")
            time.sleep(60)
            continue

        delay = random.uniform(20, 40)
        log.info(f"다음 확인까지 {delay:.1f}초 대기")
        time.sleep(delay)


if __name__ == "__main__":
    main()

File: android/app/src/main/java/com/ticketchecker/checker/InterparkChecker.kt
package com.ticketchecker.checker

import android.util.Log
import com.ticketchecker.model.InterparkTarget
import okhttp3.HttpUrl.Companion.toHttpUrl
import okhttp3.OkHttpClient
import okhttp3.Request
import org.xmlpull.v1.XmlPullParser
import org.xmlpull.v1.XmlPullParserFactory
import java.io.StringReader

data class InterparkSeatGrade(
    val seatGradeName: String,
    val remainCnt: Int,
    val salesPrice: String
)

sealed class InterparkCheckResult {
    data class Available(val grades: List<InterparkSeatGrade>) : InterparkCheckResult()
    object NoTickets : InterparkCheckResult()
    data class Error(val message: String) : InterparkCheckResult()
    object SessionExpired : InterparkCheckResult()
}

class InterparkChecker(private val client: OkHttpClient) {

    companion object {
        private const val TAG = "InterparkChecker"
        private const val API_URL = "https://poticket.interpark.com/Book/Lib/BookInfoXml.asp"
        private const val USER_AGENT =
            "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/147.0.0.0 Safari/537.36"
    }

    fun check(target: InterparkTarget): InterparkCheckResult {
        return try {
            val cookieHeader = target.cookies.entries.joinToString("; ") { "${it.key}=${it.value}" }

            val urlBuilder = API_URL.toHttpUrl().newBuilder()
                .addQueryParameter("Flag", "OrderSeatGrade")
                .addQueryParameter("GoodsCode", target.goodsCode)
                .addQueryParameter("PlaceCode", target.placeCode)
                .addQueryParameter("BizCode", target.bizCode)
                .addQueryParameter("PlaySeq", target.playSeq)
                .addQueryParameter("SessionId", target.sessionId)
                .addQueryParameter("InterlockingGoods", "")
                .build()

            val request = Request.Builder()
                .url(urlBuilder)
                .get()
                .header("User-Agent", USER_AGENT)
                .header("Accept", "application/xml, text/xml, */*")
                .header("Accept-Language", "ko-KR,ko;q=0.9")
                .header("Referer", "https://poticket.interpark.com/")
                .apply {
                    if (cookieHeader.isNotEmpty()) {
                        header("Cookie", cookieHeader)
                    }
                }
                .build()

            val response = client.newCall(request).execute()
            val body = response.body?.string() ?: return InterparkCheckResult.Error("Empty response")

            Log.d(TAG, "Response code: ${response.code}, body length: ${body.length}")

            if (!response.isSuccessful) {
                if (response.code == 401 || response.code == 403) {
                    return InterparkCheckResult.SessionExpired
                }
                return InterparkCheckResult.Error("HTTP ${response.code}")
            }

            if (body.contains("로그인", ignoreCase = true) && body.contains("session", ignoreCase = true)) {
                return InterparkCheckResult.SessionExpired
            }

            parseXmlResponse(body, target.watchGrades)
        } catch (e: Exception) {
            Log.e(TAG, "Check failed", e)
            InterparkCheckResult.Error(e.message ?: "Unknown error")
        }
    }

    /** API를 호출해 전체 좌석 등급명 목록을 반환합니다 (RemainCnt 무관). */
    fun fetchAllGradeNames(target: InterparkTarget): List<String> {
        return try {
            val cookieHeader = target.cookies.entries.joinToString("; ") { "${it.key}=${it.value}" }
            val url = API_URL.toHttpUrl().newBuilder()
                .addQueryParameter("Flag", "OrderSeatGrade")
                .addQueryParameter("GoodsCode", target.goodsCode)
                .addQueryParameter("PlaceCode", target.placeCode)
                .addQueryParameter("BizCode", target.bizCode)
                .addQueryParameter("PlaySeq", target.playSeq)
                .addQueryParameter("SessionId", target.sessionId)
                .addQueryParameter("InterlockingGoods", "")
                .build()

            val request = Request.Builder()
                .url(url)
                .get()
                .header("User-Agent", USER_AGENT)
                .header("Accept", "application/xml, text/xml, */*")
                .header("Accept-Language", "ko-KR,ko;q=0.9")
                .header("Referer", "https://poticket.interpark.com/")
                .apply { if (cookieHeader.isNotEmpty()) header("Cookie", cookieHeader) }
                .build()

            val response = client.newCall(request).execute()
            if (!response.isSuccessful) return emptyList()
            val body = response.body?.string() ?: return emptyList()
            parseAllGradeNames(body)
        } catch (e: Exception) {
            Log.e(TAG, "fetchAllGradeNames failed", e)
            emptyList()
        }
    }

    private fun parseAllGradeNames(xml: String): List<String> {
        return try {
            val factory = XmlPullParserFactory.newInstance()
            val parser = factory.newPullParser()
            parser.setInput(StringReader(xml))

            val names = mutableListOf<String>()
            var inTable = false
            var currentTag = ""

            var eventType = parser.eventType
            while (eventType != XmlPullParser.END_DOCUMENT) {
                when (eventType) {
                    XmlPullParser.START_TAG -> {
                        currentTag = parser.name
                        if (currentTag == "Table") inTable = true
                    }
                    XmlPullParser.TEXT -> {
                        if (inTable && currentTag == "SeatGradeName") {
                            val name = parser.text.trim()
                            if (name.isNotEmpty()) names.add(name)
                        }
                    }
                    XmlPullParser.END_TAG -> {
                        if (parser.name == "Table") { inTable = false; currentTag = "" }
                    }
                }
                eventType = parser.next()
            }
            names
        } catch (e: Exception) {
            Log.e(TAG, "parseAllGradeNames failed", e)
            emptyList()
        }
    }

    private fun parseXmlResponse(xml: String, watchGrades: List<String>): InterparkCheckResult {
        return try {
            val factory = XmlPullParserFactory.newInstance()
            val parser = factory.newPullParser()
            parser.setInput(StringReader(xml))

            val grades = mutableListOf<InterparkSeatGrade>()
            var inTable = false
            var seatGradeName = ""
            var remainCnt = 0
            var salesPrice = ""
            var currentTag = ""

            var eventType = parser.eventType
            while (eventType != XmlPullParser.END_DOCUMENT) {
                when (eventType) {
                    XmlPullParser.START_TAG -> {
                        currentTag = parser.name
                        if (currentTag == "Table") {
                            inTable = true
                            seatGradeName = ""
                            remainCnt = 0
                            salesPrice = ""
                        }
                    }
                    XmlPullParser.TEXT -> {
                        if (inTable) {
                            val text = parser.text.trim()
                            when (currentTag) {
                                "SeatGradeName" -> seatGradeName = text
                                "RemainCnt" -> remainCnt = text.toIntOrNull() ?: 0
                                "SalesPrice" -> salesPrice = text
                            }
                        }
                    }
                    XmlPullParser.END_TAG -> {
                        if (parser.name == "Table" && inTable) {
                            inTable = false
                            Log.d(TAG, "등급: $seatGradeName | 잔여: ${remainCnt}석 | 가격: ${salesPrice}원")
                            if (remainCnt > 0) {
                                val gradeMatch = watchGrades.isEmpty() ||
                                        watchGrades.any { it.equals(seatGradeName, ignoreCase = true) }
                                if (gradeMatch) {
                                    grades.add(
                                        InterparkSeatGrade(
                                            seatGradeName = seatGradeName,
                                            remainCnt = remainCnt,
                                            salesPrice = salesPrice
                                        )
                                    )
                                }
                            }
                            currentTag = ""
                        }
                    }
                }
                eventType = parser.next()
            }

            if (grades.isNotEmpty()) {
                InterparkCheckResult.Available(grades)
            } else {
                InterparkCheckResult.NoTickets
            }
        } catch (e: Exception) {
            Log.e(TAG, "XML parse failed", e)
            InterparkCheckResult.Error("XML parse error: ${e.message}")
        }
    }
}
